@JsonApiRelationshipRepository and @Transactional in Spring Boot

Hi there! It seems like it is impossible to use Katharsis in a transactional manner in a Spring Boot app.
When I add `@Transactional` annotation to a class annotated with `@JsonApiRelationshipRepository` or to a method within such a class I get NPE:

> io.katharsis.invoker.KatharsisInvokerException: java.lang.NullPointerException
	at io.katharsis.spring.KatharsisFilterV2.invoke(KatharsisFilterV2.java:104) ~[katharsis-spring-2.3.0.jar:na]
	at io.katharsis.spring.KatharsisFilterV2.doFilter(KatharsisFilterV2.java:81) ~[katharsis-spring-2.3.0.jar:na]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) [spring-web-4.3.0.BUILD-SNAPSHOT.jar:4.3.0.BUILD-SNAPSHOT]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.0.BUILD-SNAPSHOT.jar:4.3.0.BUILD-SNAPSHOT]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [tomcat-embed-core-8.0.32.jar:8.0.32]
	...
Caused by: java.lang.NullPointerException: null
	at io.katharsis.resource.registry.repository.AnnotatedRelationshipEntryBuilder.getTargetAffiliation(AnnotatedRelationshipEntryBuilder.java:21) ~[katharsis-core-2.3.0.jar:na]
	at io.katharsis.resource.registry.RegistryEntry.getRelationshipRepositoryForClass(RegistryEntry.java:60) ~[katharsis-core-2.3.0.jar:na]
	at io.katharsis.dispatcher.controller.resource.RelationshipsResourceUpsert.handle(RelationshipsResourceUpsert.java:107) ~[katharsis-core-2.3.0.jar:na]
	at io.katharsis.dispatcher.RequestDispatcher.dispatchRequest(RequestDispatcher.java:46) ~[katharsis-core-2.3.0.jar:na]
	at io.katharsis.spring.KatharsisFilterV2.dispatchRequest(KatharsisFilterV2.java:128) ~[katharsis-spring-2.3.0.jar:na]
	at io.katharsis.spring.KatharsisFilterV2.invoke(KatharsisFilterV2.java:102) ~[katharsis-spring-2.3.0.jar:na]
	... 34 common frames omitted

I dug into the problem a little bit. It seems like Spring is creating an additional Proxy over the repo which has the only annotation (`@Transactional`). Katharsis just doesn't find the `@JsonApiRelationshipRepository` which is needed to get target class. 

Any idea how this issue can be solved? Thanks!

--- a/src/main/java/io/katharsis/resource/registry/repository/AnnotatedRelationshipEntryBuilder.java
+++ b/src/main/java/io/katharsis/resource/registry/repository/AnnotatedRelationshipEntryBuilder.java
@@ -5,6 +5,8 @@ import io.katharsis.repository.RelationshipRepository;
 import io.katharsis.repository.RepositoryMethodParameterProvider;
 import io.katharsis.repository.adapter.RelationshipRepositoryAdapter;
 import io.katharsis.repository.annotations.JsonApiRelationshipRepository;
+import io.katharsis.utils.ClassUtils;
+import io.katharsis.utils.java.Optional;
 
 public class AnnotatedRelationshipEntryBuilder<T, D> implements RelationshipEntry<T, D> {
 
@@ -16,9 +18,21 @@ public class AnnotatedRelationshipEntryBuilder<T, D> implements RelationshipEntr
 
     @Override
     public Class<?> getTargetAffiliation() {
-        return repositoryInstance.getClass()
-            .getAnnotation(JsonApiRelationshipRepository.class)
-            .target();
+        final Optional<JsonApiRelationshipRepository> annotation = ClassUtils.getAnnotation(
+                repositoryInstance.getClass(),
+                JsonApiRelationshipRepository.class
+        );
+
+        if (annotation.isPresent()) {
+            return annotation.get().target();
+        } else {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Class %s must be annotated with @JsonApiRelationshipRepository",
+                            repositoryInstance.getClass().getName()
+                    )
+            );
+        }
     }
 
     public RelationshipRepository<T, ?, ?, ?> build(RepositoryMethodParameterProvider parameterProvider) {
@@ -28,7 +42,7 @@ public class AnnotatedRelationshipEntryBuilder<T, D> implements RelationshipEntr
     @Override
     public String toString() {
         return "AnnotatedRelationshipEntryBuilder{" +
-            "repositoryInstance=" + repositoryInstance +
-            '}';
+                "repositoryInstance=" + repositoryInstance +
+                '}';
     }
 }

--- a/src/test/java/io/katharsis/resource/registry/repository/AnnotatedRelationshipEntryBuilderTest.java
+++ b/src/test/java/io/katharsis/resource/registry/repository/AnnotatedRelationshipEntryBuilderTest.java
@@ -1,0 +1,62 @@
+package io.katharsis.resource.registry.repository;
+
+import io.katharsis.repository.annotations.JsonApiRelationshipRepository;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class AnnotatedRelationshipEntryBuilderTest {
+
+    @Test
+    public void onInstanceOfAnnotatedRelationshipRepositoryShouldReturnTargetClass() {
+
+        // GIVEN
+        final SimpleRelationshipRepository repositoryInstance = new SimpleRelationshipRepository();
+        final AnnotatedRelationshipEntryBuilder builder = new AnnotatedRelationshipEntryBuilder(repositoryInstance);
+
+
+        // WHEN
+        final Class<?> targetClass = builder.getTargetAffiliation();
+
+        // THEN
+        assertThat(targetClass).isEqualTo(Character.class);
+    }
+
+    @Test
+    public void onInstanceOfAnonymousDescendantOfAnnotatedRelationshipRepositoryShouldReturnTargetClass() {
+
+        // GIVEN
+        final Object repositoryInstance = new SimpleRelationshipRepository() {};
+        final AnnotatedRelationshipEntryBuilder builder = new AnnotatedRelationshipEntryBuilder(repositoryInstance);
+
+        // WHEN
+        final Class<?> targetClass = builder.getTargetAffiliation();
+
+        // THEN
+        assertThat(targetClass).isEqualTo(Character.class);
+
+    }
+
+    @Test
+    public void onInstanceOfNonAnnotatedClassShouldThrowIllegalArgumentException() {
+
+        // GIVEN
+        final Object repositoryInstance = new Object();
+        final AnnotatedRelationshipEntryBuilder builder = new AnnotatedRelationshipEntryBuilder(repositoryInstance);
+
+        // WHEN
+        try {
+            builder.getTargetAffiliation();
+        } catch (Exception e) {
+            // THEN
+            assertThat(e).isInstanceOf(IllegalArgumentException.class);
+        }
+    }
+
+    @JsonApiRelationshipRepository(source = String.class, target = Character.class)
+    public static class SimpleRelationshipRepository {
+
+    }
+
+}

